txnsync/outgoing.go

```diff
@@ -0,0 +1,298 @@
+func (encoder *messageAsyncEncoder) asyncEncodeAndSend(interface{}) interface{} {
+	defer encoder.state.messageSendWaitGroup.Done()
+
+	var err error
+	if len(encoder.messageData.transactionGroups) > 0 {
+		encoder.messageData.message.TransactionGroups, err = encoder.state.encodeTransactionGroups(encoder.messageData.transactionGroups, encoder.peerDataExchangeRate)
+		if err != nil {
+			encoder.state.log.Warnf("unable to encode transaction groups : %v", err)
+		}
+		encoder.messageData.transactionGroups = nil // clear out to allow GC to reclaim
+	}
+
+	encodedMessage := encoder.messageData.message.MarshalMsg(getMessageBuffer())
+	encoder.messageData.encodedMessageSize = len(encodedMessage)
+	// now that the message is ready, we can discard the encoded transaction group slice to allow the GC to collect it.
+	releaseEncodedTransactionGroups(encoder.messageData.message.TransactionGroups.Bytes)
+	// record the timestamp here, before sending the raw bytes to the network :
+	// the time we spend on the network package might include the network processing time, which
+	// we want to make sure we avoid.
+	encoder.messageData.sentTimestamp = encoder.roundClock.Since()
+
+	encoder.state.node.SendPeerMessage(encoder.messageData.peer.networkPeer, encodedMessage, encoder.asyncMessageSent)
+	releaseMessageBuffer(encodedMessage)
+
+	encoder.messageData.message.TransactionGroups.Bytes = nil
+	// increase the metric for total messages sent.
+	txsyncOutgoingMessagesTotal.Inc(nil)
+	return nil
+}
+
+// enqueue add the given message encoding task to the execution pool, and increase the waitgroup as needed.
+func (encoder *messageAsyncEncoder) enqueue() {
+	encoder.state.messageSendWaitGroup.Add(1)
+	if err := encoder.state.threadpool.EnqueueBacklog(context.Background(), encoder.asyncEncodeAndSend, nil, nil); err != nil {
+		encoder.state.messageSendWaitGroup.Done()
+	}
+}
+
+// pendingTransactionGroupsSnapshot is used to represent a snapshot of a pending transaction groups along with the latestLocallyOriginatedGroupCounter value.
+// The goal is to ensure we're "capturing"  this only once per `sendMessageLoop` call. In order to do so, we allocate that structure on the stack, and passing
+// a pointer to that structure downstream.
+type pendingTransactionGroupsSnapshot struct {
+	pendingTransactionsGroups           []transactions.SignedTxGroup
+	latestLocallyOriginatedGroupCounter uint64
+}
+
+func (s *syncState) sendMessageLoop(currentTime time.Duration, deadline timers.DeadlineMonitor, peers []*Peer) {
+	if len(peers) == 0 {
+		// no peers - no messages that need to be sent.
+		return
+	}
+	var pendingTransactions pendingTransactionGroupsSnapshot
+	profGetTxnsGroups := s.profiler.getElement(profElementGetTxnsGroups)
+	profAssembleMessage := s.profiler.getElement(profElementAssembleMessage)
+	profGetTxnsGroups.start()
+	pendingTransactions.pendingTransactionsGroups, pendingTransactions.latestLocallyOriginatedGroupCounter = s.node.GetPendingTransactionGroups()
```

Comment:
that wouldn't be a bad idea - but we would need to make sure it's sufficiently optimized so that we don't perform the same optimization on the same input dataset twice.